library/Garden/TwigTrait.php

```diff
@@ -0,0 +1,25 @@
+<?php
+/**
+ * @author Alexander Kim <alexander.k@vanillaforums.com>
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license GPLv2
+ */
+
+namespace Garden;
+
+/**
+ * For classes that need to cache some static values and configs.
+ *
+ */
+trait TwigTrait {
+    protected static $defaultFolder = PATH_ROOT.'/plugins/knowledge/views';
+    /**
+     * Twig templating environment initialize
+     *
+     * @return \Twig_Environment
+     */
+    protected static function twigInit() {
```

Comment:
Is it possible to make this something that doesn't need to be instantiated and configured by every controller that uses this trait? It'd be great if we had a single Twig instance that could be used throughout Vanilla and not one per class.

Reply:
That is something I feel we have opposite points of view.
I see that practice everywhere in old Vanilla code sending same objects/containers everywhere.
Why Don't you like to have very well isolated areas? 
In this case whenever I use Twig in my controller class I will be absolutely confident that it is my fresh object.
Why should we reuse this object in any other class? what is the profit/pros? I see only cons.
By the way, we can implement singleton to have shared instance using same trait in future. I just can't get why do we need it.

Reply:
It makes a lot more sense to me to configure the view layer one time, and allow the application to use it in as many areas as it needs. For example, if an addon (or Vanilla) wanted to extend Twig, it shouldn't need to be done more than once. We could use a shared instance to achieve that. With a per-class instance, it becomes more difficult. Are there any real "pros" to potentially having multiple instances of Twig, each with their own separate config ("because they're isolated" isn't specific enough)? The only real "con" I can see from a shared instance would be possible data pollution, but I don't think that's even a concern here.

Reply:
>For example, if an addon (or Vanilla) wanted to extend Twig, it shouldn't need to be done more than once.
I can't get this idea. If we want to allow to extend - that means to extend as many time as many plugins we have. That is kind definition of OOP "extends" term I have in my mind. So when I see "it shouldn't need to be done more than once" following after "wanted to extend" I lose the idea completely. 

>With a per-class instance, it becomes more difficult
IMHO it becomes more clear, inheritance is the thing. Whenever I am in my child class - I can easily use any parent methods and I can redefine any or create something new when I need. So I can't get what exactly you worry about. probably we need to discuss some example to illustrate the issue/problem you want to avoid

Reply:
@initvector If we wanted to do some global configuration of twig configuration couldn't we do it in this trait?

I prefer taking the magic out of the view rendering and making it more explicit in this way. We currently do not have any requirement or plan or intention to require any weird lookup chains for where the template lives, so:

If the declaration of what view we need to render is explicit, why can't the declaration of your templating system?

Reply:
> If we wanted to do some global configuration of twig configuration couldn't we do it in this trait?

Sure! I'm not opposed to that. Ultimately, my primary concern is that Vanilla's usage of Twig won't be this simple for very long. I anticipate needing to configure filters or tags or extensions or helpers or whatever else. I don't see the real benefit in having to create a new instance and potentially reconfigure it with these items, every time something needs to be rendered. I don't think the controller necessarily needs to be concerned with defining the details of the view layer. There doesn't seem to be a persistent state for a Twig instance, outside of its configuration, so I'm not seeing the benefit of having multiple instances with separate configurations for the sake of potentially shaving off some bits or ticks in certain scenarios. **However**, this is a team decision. If I'm in the minority on this, that's fine. I won't consider this to be a blocker.